hw/ip/prim/dv/prim_ram_scr/cpp/scramble_model.cc

```diff
@@ -0,0 +1,373 @@
+      } else {
+        // Other bits are taken from nonce. Each PRINCE instaniations will used
+        // different nonce bits.
+        or_vector_bit(
+            iv, j,
+            read_vector_bit(
+                nonce, (j - addr_width) + i * (kPrinceWidth - addr_width)));
+      }
+    }
+
+    // PRINCE C reference model works on big-endian byte order
+    iv = byte_swap_vector(iv);
+    std::vector<uint8_t> key_be = byte_swap_vector(key);
+
+    // Apply PRINCE to IV to produce keystream
+    std::vector<uint8_t> keystream_block(kPrinceWidthByte);
+    prince_enc_dec(&iv[0], &key_be[0], &keystream_block[0], 0, num_half_rounds,
+                   0);
+
+    // Flip keystream into little endian order and add to keystream vector
+    keystream_block = byte_swap_vector(keystream_block);
+    keystream.insert(keystream.end(), keystream_block.begin(),
+                     keystream_block.end());
+  }
+
+  return keystream;
+}
+
+// Split incoming data into `subst_perm_width` chunks and individually apply the
+// substitution/permutation layer to each
+static std::vector<uint8_t> scramble_subst_perm_full_width(
+    const std::vector<uint8_t> &in, uint32_t bit_width,
+    uint32_t subst_perm_width, bool enc) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+
+  // Determine how many bytes each `subst_perm_width` chunk is and how many
+  // chunks are needed to cover the full `bit_width`.
+  uint32_t subst_perm_bytes = (subst_perm_width + 7) / 8;
+  uint32_t subst_perm_blocks =
+      (bit_width + subst_perm_width - 1) / subst_perm_width;
+
+  std::vector<uint8_t> out(in.size(), 0);
+  std::vector<uint8_t> zero_key(subst_perm_bytes, 0);
+
+  for (int i = 0; i < subst_perm_blocks; ++i) {
+    uint32_t local_subst_perm_width;
+
+    // Where `bit_width` does not evenly divide into `subst_perm_width` the
+    // final chunk is smaller.
+    if (i == subst_perm_blocks - 1) {
+      local_subst_perm_width =
+          bit_width - subst_perm_width * (subst_perm_blocks - 1);
+    } else {
+      local_subst_perm_width = subst_perm_width;
+    }
+
+    std::vector<uint8_t> subst_perm_data(subst_perm_bytes, 0);
+
+    // Extract bits from `in` for this chunk
+    for (int j = 0; j < local_subst_perm_width; ++j) {
+      or_vector_bit(subst_perm_data, j,
+                    read_vector_bit(in, j + i * subst_perm_width));
+    }
+
+    // Apply the subtitution/permutation layer to the chunk
+    std::vector<uint8_t> subst_perm_out;
+    if (enc) {
+      subst_perm_out = scramble_subst_perm_enc(subst_perm_data, zero_key,
+                                               local_subst_perm_width,
+                                               kNumDataSubstPermRounds);
+    } else {
+      subst_perm_out = scramble_subst_perm_dec(subst_perm_data, zero_key,
+                                               local_subst_perm_width,
+                                               kNumDataSubstPermRounds);
+    }
```

Comment:
I'd consider hoisting this decision out of the loop and assigning to a "scrambler" function pointer. Since this is C++11, you don't even need to faff around with the types!
```c++
auto sp_scrambler = enc ? scramble_subst_perm_enc : scramble_subst_perm_dec;
```

Reply:
I was intentionally avoiding C++11 and onwards features as I thought VCS required some ancient compiler, or does that give C++11 support?

Reply:
We use C++11 heavily and you can rely on it (the OTBN code even uses std::regex, which was pretty flaky with early GCC support)

Reply:
Switched, also added use of `auto` in various places as there's no need to avoid C++11 features.